src/ode_solver/low_storage_runge_kutta_ode_solver.cpp

```diff
@@ -0,0 +1,188 @@
+#include "low_storage_runge_kutta_ode_solver.h"
+
+namespace PHiLiP {
+namespace ODE {
+
+template <int dim, typename real, int n_rk_stages, typename MeshType> 
+LowStorageRungeKuttaODESolver<dim,real,n_rk_stages, MeshType>::LowStorageRungeKuttaODESolver(std::shared_ptr< DGBase<dim, real, MeshType> > dg_input,
+        std::shared_ptr<LowStorageRKTableauBase<dim,real,MeshType>> rk_tableau_input,
+        std::shared_ptr<EmptyRRKBase<dim,real,MeshType>> RRK_object_input)
+        : ODESolverBase<dim,real,MeshType>(dg_input)
+        , butcher_tableau(rk_tableau_input)
+        , relaxation_runge_kutta(RRK_object_input)
+        , solver(dg_input)
+{       epsilon[0] = 1.0;
+        epsilon[1] = 1.0;
+        epsilon[2] = 1.0; 
+}
+
+template <int dim, typename real, int n_rk_stages, typename MeshType>
+void LowStorageRungeKuttaODESolver<dim,real,n_rk_stages, MeshType>::step_in_time (real dt, const bool pseudotime)
+{
+    this->original_time_step = dt;
+    this->solution_update = this->dg->solution; //storing u_n
+    (void) pseudotime;
+
+    dealii::LinearAlgebra::distributed::Vector<double> storage_register_2;
+    storage_register_2.reinit(this->solution_update);
+    storage_register_2*=0;
+    dealii::LinearAlgebra::distributed::Vector<double> storage_register_1;
+    storage_register_1.reinit(this->solution_update);
+    storage_register_1 = this->dg->solution;
+    dealii::LinearAlgebra::distributed::Vector<double> storage_register_3 = storage_register_1;
+
+    dealii::LinearAlgebra::distributed::Vector<double> rhs = storage_register_1;
+
+    dealii::LinearAlgebra::distributed::Vector<double> storage_register_4;
+
+    if (this->butcher_tableau->get_b_hat(0) != 0.0){
+        storage_register_4 = storage_register_1;
+    } 
+    
+    double sum_delta = 0;
+
+    //double atol = 0.001;
+    //double rtol = 0.001;
+    double error = 0.0;
+    w = 0.0;
+
+    for (int i = 1; i < n_rk_stages +1; i++ ){
+        // storage_register_2 = storage_register_2 + delta[i-1] * storage_register_1;
+        storage_register_2.add(this->butcher_tableau->get_delta(i-1) , storage_register_1);
+        this->dg->solution = rhs;
```

Comment:
I think this might be possible to do without the extra vector `rhs`, and instead using `storage_register_1`. I haven't gone through the loop in detail, but maybe you could check that.

Reply:
I was looking into this but I think it would be more complicated using storage_register_1 instead of rhs because in that step to calculate storage_register_1, it also uses the unchanged value of storage_register_1. I looked through the Julia code that the paper references and they also use another variable to store f(storage_register_1)

Reply:
Okay! Thanks for checking their code.